islpy/oppool.py

```diff
@@ -0,0 +1,830 @@
+            old_name = obj.get_dim_name(dt, pos) or new_name
+            new_obj = new_obj.set_dim_name(dt, pos, new_name)
+            lift_map[new_name] = old_name
+
+    return new_obj, lift_map
+
+
+def normalize_binary_isl_obj(obj1: "NormalizedISLObj",
+                             obj2: "NormalizedISLObj") -> Tuple["NormalizedISLObj",
+                                                                "NormalizedISLObj"]:
+    vng = UniqueNameGenerator(forced_prefix="_islpy")
+    lift_map1 = {}
+    lift_map2 = {}
+    unlift_map1 = {}
+
+    # FIXME: What if these liftings turn out to be expensive?
+    new_obj1 = obj1.lift()
+    new_obj2 = obj2.lift()
+
+    for old_name, (dt, pos) in new_obj1.get_var_dict().items():
+        new_name = _gen_name(vng, dt)
+        if not isinstance(new_obj1, isl.PwAff):
+            new_obj1 = new_obj1.set_dim_name(dt, pos, new_name)
+        else:
+            new_obj1 = new_obj1.set_dim_id(dt, pos,
+                                           isl.Id.read_from_str(new_obj1.get_ctx(),
+                                                                new_name))
+        lift_map1[new_name] = old_name
+        unlift_map1[old_name] = new_name
+
+    for old_name, (dt, pos) in new_obj2.get_var_dict().items():
+        new_name = unlift_map1.get(old_name) or _gen_name(vng, dt)
+        if not isinstance(new_obj2, isl.PwAff):
+            new_obj2 = new_obj2.set_dim_name(dt, pos, new_name)
+        else:
+            new_obj2 = new_obj2.set_dim_id(dt, pos,
+                                           isl.Id.read_from_str(new_obj2.get_ctx(),
+                                                                new_name))
+        lift_map2[new_name] = old_name
+
+    return (obj1.copy(new_obj1, lift_map1),
+            obj2.copy(new_obj2, lift_map2))
+
+
+def normalize_based_on_template(obj: "NormalizedISLObj",
+                                tmpl: "NormalizedISLObj") -> "NormalizedISLObj":
+    """
+    Like :func:`normalize_binary_isl_obj`, but obj1 does not support set_dim_name.
+    """
+    assert set(obj.lift_map.values()) == set(tmpl.lift_map.values())
+    obj_var_dict = obj.ground_obj.get_var_dict()
+    new_ground = obj.ground_obj
+    for name_in_tmplt_grnd, name in tmpl.lift_map.items():
+        name_in_obj_grnd = obj.unlift_map[name]
+        dt, pos = obj_var_dict[name_in_obj_grnd]
+        new_ground = new_ground.set_dim_name(dt, pos, name_in_tmplt_grnd)
+
+    return obj.copy(new_ground, tmpl.lift_map)
+
+
+@dataclass
+class ISLOpMemoizer:
+    """
+    A memoizer for ISL operations.
+    """
+    cache: Dict[Any, Any] = field(default_factory=dict)
+
+    def __call__(self, f, args: Tuple[Any, ...]):
+        try:
+            return self.cache[(str(f), args)]
+        except KeyError:
+            result = f(*args)
+            self.cache[(str(f), args)] = result
+            return result
+
+
+@dataclass
+class NormalizedISLObj:
+    ground_obj: BaseType
+    lift_map: Dict[str, str]
```

Comment:
Could you document what keys and values are here? I'm particularly worried about the isl end being a `str`, because that can't be cheaply turned into what isl talks about. I was expecting a `(dim_type, dim)` tuple here.